File: CPP/program/Garage.cpp
// Menggunakan pragma once agar header file hanya di-include satu kali
#pragma once

// Memasukan library yang digunakan, serta kelas yang dimiliki oleh Car dan Motorcycle yang merupakan composition
#include <iostream>
#include <string>
#include <vector>
#include "Car.cpp"
#include "Motorcycle.cpp"

// Deklarasi namespace
using namespace std;

/*
    Kelas Garage yang digunakan untuk merepresentasikan data Garage yang merupakan Garasi tempat parkir dalam ruangan
*/
class Garage
{
    private:
        // Atribut yang digunakan pada kelas Garage terdiri dari nama dan luas dari garage tersebut. Garage ini memiliki daftar Mobil dan daftar Motor
        string nama;
        string luas;
        vector<Car*> daftarMobil;
        vector<Motorcycle*> daftarMotor;
        
    public:
        // Constructor
        Garage()
        {

        }

        // Constructor : Overloading
        Garage(string nama)
        {  
            this->nama = nama;
            luas = "";
        }

        // Constructor : Overloading
        Garage(string nama, string luas)
        {  
            this->nama = nama;
            this->luas = luas;
        }

        // Method setter untuk set setiap nilai atribut pada kelas Garage
        void set_nama(string nama)
        {
            this->nama = nama;
        }

        void set_luas(string luas)
        {
            this->luas = luas;
        }

        // Method untuk menambahkan motor ke dalam daftar motor di dalam class ini
        void add_motor(Motorcycle* motor) {
            daftarMotor.push_back(motor);
        }

        // Method untuk menambahkan mobil ke dalam daftar mobil di dalam class ini
        void add_mobil(Car* mobil) {
            daftarMobil.push_back(mobil);
        }

        // Method getter untuk mendapatkan nilai setiap atribut pada kelas Garage
        string get_nama()
        {
            return nama;
        }

        string get_luas()
        {
            return luas;
        }

        // Digunakan untuk menampilkan isi daftar kendaraan dari si garasi
        void get_datakendaraan() {
            cout << "Daftar Mobil di Garasi" << nama << ":" << endl << endl;
            for (Car* mobil : daftarMobil) {
                cout << "Plat Mobil " << mobil->get_platnomor() << " dengan Merk " << mobil->get_merk() << 
                " dengan Jumlah Pintu " << mobil->get_jumlahpintu() << " dengan Jumlah Kursi " << mobil->get_jumlahkursi()
                << " dirilis pada Tahun " << mobil->get_tahunproduksi() << endl;
            }
            cout << "-----------------------------------------------------------------" << endl;
            cout << "Daftar Motor di Garasi" << nama << ":" << endl << endl;;
            for (Motorcycle* motor : daftarMotor) {
                cout << "Plat Motor " << motor->get_platnomor() << " dengan Merk " << motor->get_merk() << 
                " dengan Jenis Motor " << motor->get_jenismotor() << " dengan Kapasitas Tangki " << motor->get_kapasitastangki()
                << " dirilis pada Tahun " << motor->get_tahunproduksi() << endl;
            }
        }

        // Destructor
        ~Garage()
        {

        }
};

File: Python/program/Garage.py
# Deklarasi Kelas Garasi
class Garage:
    __namagarasi=""
    __luasgarasi=0
    __daftarkendaraanMotor=[]
    __daftarkendaraanMobil=[]

    # Konstruktor kelas Garasi
    def __init__(self, namagarasi="", luasgarasi=0):
        self.__namagarasi = namagarasi
        self.__luasgarasi = luasgarasi
        self.__daftarkendaraanMotor=[]
        self.__daftarkendaraanMobil=[]

    # Getter dan setter nama garasi
    def get_namagarasi(self):
        return self.__namagarasi
    def set_namagarasi(self, namagarasi):
        self.__namagarasi = namagarasi

    # Getter dan setter luas garasi
    def get_luasgarasi(self):
        return self.__luasgarasi
    def set_luasgarasi(self, luasgarasi):
        self.__luasgarasi = luasgarasi

    # Setter untuk menambah daftar kendaraan motor dan mobil
    def add_motor(self, motor):
        self.__daftarkendaraanMotor.append(motor)
    def add_mobil(self, mobil):
        self.__daftarkendaraanMobil.append(mobil)
    
    # Getter untuk menampilkan daftar kendaraan dalam garasi
    def display_kendaraan(self):
        print(f"======================================================================================================================================================")
        print(f"! Garasi {self.__namagarasi} dengan luas {self.__luasgarasi} m2 !\n")
        print(f"Daftar Mobil:")
        for mobil in self.__daftarkendaraanMobil:
            print(f"- Mobil dengan plat nomor {mobil.get_platnomor()} merk {mobil.get_merk()} diproduksi pada tahun {mobil.get_tahunproduksi()} berwarna {mobil.get_warna()} dengan jumlah kursi {mobil.get_jumlahkursi()} dan jumlah pintu {mobil.get_jumlahpintu()}")
        
        print(f"\nDaftar Motor:")
        for motor in self.__daftarkendaraanMotor:
            print(f"- Motor {motor.get_jenismotor()} dengan plat nomor {motor.get_platnomor()} merk {motor.get_merk()} diproduksi pada tahun {motor.get_tahunproduksi()} berwarna {motor.get_warna()} ber kapasitas {motor.get_kapasitastangki()} liter")
        print(f"======================================================================================================================================================\n")